### ISSUE TITLE
Piece.link creates a v1 Piece CID if link property not set

### ISSUE THREAD

**Issue by olizilla** [storacha/data-segment#30]
https://github.com/web3-storage/data-segment/blob/4ad3d34cb4b4144e4e3e899fe9f8d72564c0cad5/src/piece.js#L169-L173

Is that expected? `.link` returns a V2 Piece CID if you create it via `Piece.fromPayload`.

More generally It'd be handy if this library provided a utilities to convert between a V1 and a V2 Piece CID.

**Comment by Gozala** [comment]
> Is that expected? .link returns a V2 Piece CID if you create it via Piece.fromPayload.

NO it should return V1 Piece CID, if it returns V2 that is a bug although looking at the code I can't see how that could be a case given that both digest code and format code are for the V1 there. Was that a typo or did you actually mean that you get V1 CIDs here ?

> More generally It'd be handy if this library provided a utilities to convert between a V1 and a V2 Piece CID.

It is not possible to convert V1 → V2 because V2 has more information, that said there are utilities to do transformation between two specifically you can get PieceView from payload via `Piece.fromPayload`

https://github.com/web3-storage/data-segment/blob/4ad3d34cb4b4144e4e3e899fe9f8d72564c0cad5/src/piece.js#L33-L36


Which has [(Piece V2) link field](https://github.com/web3-storage/data-segment/blob/4ad3d34cb4b4144e4e3e899fe9f8d72564c0cad5/src/api.ts#L110) if you do want to go to Piece V1 link, you can call [`toInfo()` of the view ](https://github.com/web3-storage/data-segment/blob/4ad3d34cb4b4144e4e3e899fe9f8d72564c0cad5/src/api.ts#L115) which returns `PieceInfoView` that has V1 CID link field

https://github.com/web3-storage/data-segment/blob/4ad3d34cb4b4144e4e3e899fe9f8d72564c0cad5/src/api.ts#L204-L208


If all you have is Piece V1 CID you can not go to Piece V2 CID because you do not know the tree `height`, however if you do also know the `height` you can call `PIece.fromInfo` function

https://github.com/web3-storage/data-segment/blob/4ad3d34cb4b4144e4e3e899fe9f8d72564c0cad5/src/piece.js#L105-L114

Which will give you back `PieceInfoView` that you can pass to `Piece.toView` function that will give you backe `PieceView` which has Piece V2 link.


In other words we have `PieceView <-> PieceInfoView` conversation that is non lossy, where `PieceView` is V2 stuff and `PieceInfo` is V1 stuff, and both implement `Piece` interface that is actual data encoded in one or the other format

https://github.com/web3-storage/data-segment/blob/4ad3d34cb4b4144e4e3e899fe9f8d72564c0cad5/src/api.ts#L97-L107

**Comment by olizilla** [comment]
Got it. No bug here. Just needed more guidance. I confused the `Info` class with the `Piece` class. I was accessing the link property on a Piece instance when I got the v2 Piece CID.

My only feedback is to consider renaming `Info`. It's not obvious from the library structure that we are potentially dealing with V1 or V2 Piece CIDs, but a good amount of effort has gone in to making it work nicely with either.

The `Piece` class seems to be a wrapper for extracting Piece information from a v2 Piece CID.
...while the the `Info` class a wrapper for a Piece to get a V1 Piece CID.

I'm not close enough to this to make an informed suggestion. Naïvely I would do something like have `Piece` be a data container for the properties that makes up a piece (e.g `constructor (hight, size, root)`), and `fromLink` and from `legacyLink` static helpers and a `link` method that creates and returns a v2 Piece CID, and a `legacyLink` (or `getV1PieceCid`) that returns `{ height: number, link: cid }`.

...or otherwise just make `Info` more clearly be about v1 Piece CIDs
